File: laba5/ex4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

type RegionDirectory struct {
	regions map[string]string // Хранение регионов и их центров
}

func NewRegionDirectory() *RegionDirectory {
	return &RegionDirectory{
		regions: make(map[string]string),
	}
}

func (rd *RegionDirectory) change(region, newCenter string) {
	if center, exists := rd.regions[region]; exists {
		fmt.Printf("Region %s has changed its administrative center from %s to %s\n", region, center, newCenter)
	} else {
		fmt.Printf("New region %s with administrative center %s\n", region, newCenter)
	}
	rd.regions[region] = newCenter
}

func (rd *RegionDirectory) rename(oldRegion, newRegion string) {
	if oldRegion == newRegion ||
		!rd.exists(oldRegion) ||
		rd.exists(newRegion) {
		fmt.Println("Incorrect")
		return
	}

	center := rd.regions[oldRegion]
	delete(rd.regions, oldRegion)
	rd.regions[newRegion] = center
	fmt.Printf("%s has been renamed to %s\n", oldRegion, newRegion)
}

func (rd *RegionDirectory) about(region string) {
	if center, exists := rd.regions[region]; exists {
		fmt.Printf("%s has administrative center %s\n", region, center)
	} else {
		fmt.Println("Incorrect")
	}
}

func (rd *RegionDirectory) all() {
	regions := make([]string, 0, len(rd.regions))
	for region := range rd.regions {
		regions = append(regions, region)
	}
	sort.Strings(regions)

	for _, region := range regions {
		fmt.Printf("%s - %s\n", region, rd.regions[region])
	}
}

func (rd *RegionDirectory) exists(region string) bool {
	_, ok := rd.regions[region]
	return ok
}

func main() {
	rd := NewRegionDirectory()
	scanner := bufio.NewScanner(os.Stdin)
	
	fmt.Print("Enter number of commands: ")
	scanner.Scan()
	var n int
	fmt.Sscanf(scanner.Text(), "%d", &n)

	for i := 0; i < n && scanner.Scan(); i++ {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) == 0 {
			continue
		}

		cmd := strings.ToUpper(parts[0])
		args := parts[1:]

		switch cmd {
		case "CHANGE":
			if len(args) != 2 {
				fmt.Println("Incorrect")
				continue
			}
			rd.change(args[0], args[1])

		case "RENAME":
			if len(args) != 2 {
				fmt.Println("Incorrect")
				continue
			}
			rd.rename(args[0], args[1])

		case "ABOUT":
			if len(args) != 1 {
				fmt.Println("Incorrect")
				continue
			}
			rd.about(args[0])

		case "ALL":
			rd.all()

		default:
			fmt.Println("Incorrect")
		}
	}
}

File: laba5/ex4.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <algorithm>
#include <vector>
#include <stdexcept>
#include <limits> // Добавляем для numeric_limits

using namespace std;

class RegionDirectory {
private:
    unordered_map<string, string> regions; // region -> center

public:
    void change(const string& region, const string& new_center) {
        if (regions.find(region) != regions.end()) {
            string old_center = regions[region];
            regions[region] = new_center;
            cout << "Region " << region << " has changed its administrative center from " 
                 << old_center << " to " << new_center << endl;
        } else {
            regions[region] = new_center;
            cout << "New region " << region << " with administrative center " << new_center << endl;
        }
    }

    void rename(const string& old_region, const string& new_region) {
        if (old_region == new_region || regions.count(old_region) == 0 || regions.count(new_region) > 0) {
            cerr << "Incorrect" << endl;
            return;
        }

        string center = regions[old_region];
        regions.erase(old_region);
        regions[new_region] = center;
        cout << old_region << " has been renamed to " << new_region << endl;
    }

    void about(const string& region) const {
        auto it = regions.find(region);
        if (it != regions.end()) {
            cout << region << " has administrative center " << it->second << endl;
        } else {
            cerr << "Incorrect" << endl;
        }
    }

    void all() const {
        vector<pair<string, string>> sorted_regions(regions.begin(), regions.end());
        sort(sorted_regions.begin(), sorted_regions.end());

        for (const auto& [region, center] : sorted_regions) {
            cout << region << " - " << center << endl;
        }
    }
};

string toUpper(const string& s) {
    string result = s;
    transform(result.begin(), result.end(), result.begin(), ::toupper);
    return result;
}

int main() {
    RegionDirectory directory;
    int N;
    cin >> N;
    cin.ignore();

    for (int i = 0; i < N; ++i) {
        string command;
        cin >> command;
        command = toUpper(command);

        try {
            if (command == "CHANGE") {
                string region, center;
                cin >> region >> center;
                directory.change(region, center);
            } 
            else if (command == "RENAME") {
                string old_region, new_region;
                cin >> old_region >> new_region;
                directory.rename(old_region, new_region);
            } 
            else if (command == "ABOUT") {
                string region;
                cin >> region;
                directory.about(region);
            } 
            else if (command == "ALL") {
                directory.all();
            } 
            else {
                cerr << "Incorrect" << endl;
                cin.ignore(numeric_limits<streamsize>::max(), '\n');
            }
        } catch (const exception& e) {
            cerr << "Incorrect" << endl;
        }
    }

    return 0;
}
